### DISCUSSION TITLE
Questions about mechanism of modelx that is executable with heavy recursive model.

### DISCUSSION THREAD

**Discussion by bakerwy** [fumitoh/modelx#118] [Q&A] [answered]
Hi,

I am an user who is trying various actuarial tasks with modelx

From #114,

The exported model can be executable, if it is heavy model, from python version 3.11.
I tried the previous versions and those were crashed because of stack overflow..

But in case of modelx without exported, it is executable with python version 3.10.

Is there any method or mechanism Modelx is specifically using?

I am asking this question because I want use Ray which is for advanced distributing computing.
But when I remote class to ray, it is crashed as it happened with exported model & lower python version.

I hope I can get a clue.

**Comment by fumitoh** [comment, marked as answer]
> But in case of modelx without exported, it is executable with python version 3.10.
> Is there any method or mechanism Modelx is specifically using?
Certainly! Here's a clearer and more grammatically correct version of the comment:

Yes, unlike Linux, Windows does not permit the increase of the stack memory size for the main thread of a process once it has started. To overcome this limitation, modelx, when used with Python 3.10 or older, creates an additional thread with an increased stack size specifically dedicated to the evaluation of modelx formulas.

> 
> I am asking this question because I want use Ray which is for advanced distributing computing. But when I remote class to ray, it is crashed as it happened with exported model & lower python version.

The workaround is to avoid deep recursion. Suppose you have the following formula in a modelx model and need to calculate `foo(100,000)` in the exported model.

```python
def foo(t):
    return foo(t-1) + 1 if t > 0 else 0
```

Directly calling `foo(100_000)` in the exported model leads to the deep recursion issue. Instead, executing it in segments as demonstrated below helps avoid deep recursion.

```python
for t in (range(100) + 1) * 1000:
    foo(t)
```
> 
> I hope I can get a clue.
